Add TypeScript definition

sbencoding opened the pull request:
This is a follow up PR for #21, adding `index.d.ts` and `index.d-test.ts` files.
I paid close attention to the specified [style guide](https://github.com/sindresorhus/typescript-definition-style-guide), but let me know if I've missed anything.

sbencoding committed:
Add typescript defs and test
Configure test script to check ts defs

sindresorhus commented on index.d.ts:
````diff
@@ -0,0 +1,290 @@
+		})();
+		```
+		*/
+		readonly modules?: String
````
All the `String` and `Boolean` should be lowercase. And should end in a `;`

sindresorhus commented on package.json:
```diff
@@ -42,6 +42,7 @@
 		"@types/node": "^13.7.6",
 		"ava": "^2.1.0",
 		"get-stream": "^5.1.0",
+		"tsd-check": "^0.6.0",
```
Should be `tsd` (It changed name at some point) And use latest version. The command name also changed.

sindresorhus committed:
Update index.test-d.ts

sindresorhus committed:
Update index.d.ts

sindresorhus commented on index.d.ts:
```diff
@@ -0,0 +1,305 @@
+
+		Note: Only available if the `modules` option has been set.
+
+		@default ['N/A']
```
Is this really the default? I think it should be changed to just an empty array if so.

sindresorhus commented on index.d.ts:
```diff
@@ -0,0 +1,305 @@
+
+		Note: Only available if the `apps` option has been set.
+
+		@default 'N/A'
```
This should just be `undefined` when `apps` options is not set.

sbencoding committed:
Fix tsd style, use tsd instead of tsd-check

sbencoding committed:
Update result default values

sbencoding committed:
Solve merge conflicts

sbencoding commented:
Sorry for the late reply...
I've implemented the requested changes and updated the code in a few more places, that were similar to the last 2 requested changes.

sindresorhus commented:
Hmm, this might be a problem:

> Note: It's not guaranteed that the username and windowTitle properties are returned with proper values. If they are not available, 'N/A' may be returned on English systems. In contrast, 'Nicht zutreffend' may be returned on German systems, for example.

Regarding the `N/A` change...

sbencoding committed:
Workaround for 'N/A' in different languages

sbencoding committed:
Debug Travis test fail

sbencoding committed:
Fix style errors

sbencoding committed:
Fix bug in not found text parsing

sbencoding commented:
I've implemented a workaround for the `N/A` problem. 
Basically before I execute `tasklist` with the options from the calling code, I execute `tasklist` internally, which filters the results for the PID to match 4.  
On Windows this process is always running and it's the *system* process which doesn't have a window, so the `windowTitle` parameter will always be `N/A` or equivalent in other languages.  
I don't know if this is the right way to implement this feature since this way for every call `tasklist` is executed twice, but this is the first idea that came to my mind

MarkTiedemann commented on index.js:
```diff
@@ -9,6 +9,16 @@ const csv = require('csv');
 
+function getNotFoundText() {
+	// Get output of the command
+	const stdout = childProcess.execFileSync('tasklist.exe', ['/v', '/fi', 'PID eq 4', '/nh', '/fo', 'csv']).toString();
```
A comment explaining why PID 4 was chosen would be appropriate here.

MarkTiedemann commented:
I think it would be good to avoid spawning `tasklist` twice if possible.

For example, we should not spawn twice if `verbose` is not set (`windowTitle` will only be returned if `verbose` is set).

If known system processes, such as, `System Idle Process` (PID 0), `System` (PID 4), `svchost.exe` (multiple PIDs), are part of the results, we also don't need to spawn twice.

Similarly for `/apps`, if `RuntimeBroker.exe` is part of the results, we don't need to spawn twice.

In any case, the second spawn should never be `sync` for performance reasons.

sbencoding commented:
Yeah, I agree we shouldn't run `tasklist` twice on every call, but I'm not sure right now how to do this as efficiently as possible.  

I could parse the *not found text* from the result of normal  calls where the `verbose` option is set as you've suggested, but that wouldn't work if these *"magic"* processes/apps were filtered out by the caller.
Additionally we also need to figure out the *"not found text*" when the `services` or `modules` option is used. There might be some *"magic"* values there too where we could get the *not found text* from, but I'd have to investigate.

About the `sync` problem:
If the second call, where we get the *not found text* from is `async`, there would be no problems when the `promise interface` is called. But when the stream interface is called, we would have to `await` for the *not found text* call, but that would make the `stream interface` return a promise first, and then the promise would resolve to the stream.
If we ditch the second call, where we get the *not found text*, and our *magic* process/app/whatever is not the first in the stream, than the values before the *magic* process couldn't be replaced.

sindresorhus commented:
Is there really no way to force the result to be in English? So we could just detect `N/A`.

MarkTiedemann commented:
~~`chcp` may be used: https://docs.microsoft.com/en-us/windows-server/administration/windows-commands/chcp~~

Not sure whether this has other issues, e.g. if we do change the language, do we need to change it back to the original language afterwards? Can the language always be changed to English, even on computers that don't have the English language pack installed? etc.

EDIT: No, sorry, that doesn't work. It's just changing the code page, not the language.

sindresorhus commented:
What should we do about this?

sindresorhus commented:
@sbencoding Bump :)

sindresorhus commented:
Closing for lack of response.

https://github.com/sindresorhus/tasklist/issues/28
